File: src/main/java/com/microfinance/microfinance/models/Compte.java
package com.microfinance.microfinance.models;

import jakarta.persistence.*;

@Entity
public class Compte {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    private String numeroCompte;
    private String codeBanque;
    private Double soldeActuel;
    private String typeCompte;

    @OneToOne
    @JoinColumn(name = "client_id")
    private Client client;

    public void crediter(Double montant) {
        this.soldeActuel += montant;
    }

    public void debiter(Double montant) {
        this.soldeActuel -= montant;
    }
}

File: back-end/app/Models/Compte.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\SoftDeletes;
use Illuminate\Database\Eloquent\Relations\BelongsTo;

class Compte extends Model
{
    use SoftDeletes;

    protected $fillable = [
        'client_id',
        'numero_compte',
        'code_banque',
        'solde_actuel',
        'type_compte',
        'statut',
    ];

    protected $casts = [
        'solde_actuel' => 'decimal:2',
    ];

    // -------------------------------------------------------------------------
    // Relations
    // -------------------------------------------------------------------------

    public function client(): BelongsTo
    {
        return $this->belongsTo(Client::class);
    }

    // -------------------------------------------------------------------------
    // Méthodes métier
    // -------------------------------------------------------------------------

    /**
     * @throws \Exception si solde insuffisant
     */
    public function debiter(float $montant): bool
    {
        if ($montant <= 0) {
            throw new \InvalidArgumentException('Le montant à débiter doit être positif.');
        }

        if ($this->solde_actuel < $montant) {
            throw new \Exception("Solde insuffisant. Solde actuel : {$this->solde_actuel}");
        }

        $this->decrement('solde_actuel', $montant);
        return true;
    }

    public function crediter(float $montant): bool
    {
        if ($montant <= 0) {
            throw new \InvalidArgumentException('Le montant à créditer doit être positif.');
        }

        $this->increment('solde_actuel', $montant);
        return true;
    }
}
